### PR TITLE
K8s: Improve key generation and parsing

### PR DESCRIPTION
**Pull request by toddtreece**
**What is this feature?**

* adds test coverage
* updates key format to `[/group/<group>]/resource/<resource>[/namespace/<namespace>][/name/<name>]`

**Why do we need this feature?**

This is needed to fix k8s storage and watch tests.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/apiserver/registry/generic/key.go [hunk 1]
```diff
@@ -12,63 +12,69 @@ import (
...
+		default:
+			return nil, fmt.Errorf("invalid key name: %s", key)
+		}
 	}
 
-	k.Namespace = parts[4]
-
-	if len(parts) == 5 {
-		return k, nil
+	if len(key.Resource) == 0 {
```
Comment (ryantxu) [inline pkg/apiserver/registry/generic/key.go:52]: are keys without group ever valid?

File: pkg/apiserver/registry/generic/key.go [hunk 2]
```diff
@@ -12,63 +12,69 @@ import (
...
-	k.Name = parts[5]
-
-	return k, nil
+	return key, nil
 }
 
+// String returns the string representation of the Key.
 func (k *Key) String() string {
-	s := "/" + k.Group + "/" + k.Resource
+	s := ""
```
Comment (ryantxu) [inline pkg/apiserver/registry/generic/key.go:61]: i never know if string contact vs strings.Buidler is the right approach -- do you know?

likely does not matter -- mostly for my curiosity

File: pkg/apiserver/registry/generic/key.go (hunk 1 above)
Comment (toddtreece) [reply pkg/apiserver/registry/generic/key.go:52]: yep, core v1 resources don't have a group (/api/v1/pods)

File: pkg/apiserver/registry/generic/key.go (hunk 2 above)
Comment (toddtreece) [reply pkg/apiserver/registry/generic/key.go:61]: it sounds like it depends on the length of the string. i'll benchmark

**Review by DanCech — APPROVED** [review]
kind of verbose but definitely unambiguous

File: pkg/apiserver/registry/generic/key.go (hunk 2 above)
Comment (toddtreece) [reply pkg/apiserver/registry/generic/key.go:61]:
```
BenchmarkKey_StringBuilder-32    9452661	       130.8 ns/op	     176 B/op	       4 allocs/op
BenchmarkKey_String-32    	 5975756	       196.3 ns/op	     232 B/op	       4 allocs/op
```

Comment (toddtreece) [reply pkg/apiserver/registry/generic/key.go]: updated in https://github.com/grafana/grafana/pull/90014/commits/7a0a3d72489d94efcddaec62611b9947b7b571d4

**Review by ryantxu — APPROVED** [review]
LGTM -- since the values are never stored, i don't think the length really matters

**Pull request merged**